lib/runner.js

```diff
@@ -52,11 +58,45 @@ class Runner extends Emittery {
 			if (!scheduledStart) {
 				scheduledStart = true;
 				process.nextTick(() => {
+					if (lastMetadata) {
+						const includedInRange = this.lines.some(n => n >= lastCallsiteLineNumber);
+						lastMetadata.exclusive = this.match.length === 0 ?
+							includedInRange :
+							includedInRange && lastMetadata.exclusive;
+					}
+
 					hasStarted = true;
 					this.start();
 				});
 			}
 
+			if (this.lines.length > 0) {
+				this.runOnlyExclusive = true;
+				let callsiteLineNumber;
+
+				for (const jsCallsite of callsites()) {
+					const callsite = sourceMapSupport.wrapCallSite(jsCallsite);
+					const fileName = callsite.getFileName();
+					if (fileName === this.file) {
+						callsiteLineNumber = callsite.getLineNumber();
+						break;
+					}
+				}
+
+				if (lastMetadata) {
+					// `callsiteLineNumber` is the beginning of the current test, -1 to
+					// get the last line of the previous test.
+					const range = [lastCallsiteLineNumber, callsiteLineNumber - 1];
+					const includedInRange = this.lines.some(n => n >= range[0] && n <= range[1]);
+					lastMetadata.exclusive = this.match.length === 0 ?
+						includedInRange :
+						includedInRange && lastMetadata.exclusive;
+				}
+
+				lastCallsiteLineNumber = callsiteLineNumber;
+				lastMetadata = metadata;
```

Comment:
Rather than tracking the last metedata object and callsite, perhaps you could add the callsite number to the metadata, and then adjust the test selection once all tests have been declared?